### ISSUE TITLE
trigger: combine waitForFinish and not waitForFinish behavior

### ISSUE THREAD

**Issue by MikeAlexeev** [buildbot/buildbot#3088]
Hello. 

Trigger step has option waitForFinish. If it is True, we have status update in parent build (which runs Trigger step) 
when child (which schedulled by Triggerable scheduller) finished.

But in this case parent build is blocked (not starting next step) while child is running.

On the other hand if waitForFinish is False parent build is not blocked (can start next step) but there is no status update.

I think that it would be nice to have also mode which combines this two behaviors. I.e. parent can start next step, 
but not finished until all children finished and parent updates self status respectively to children statuses.

I have hacked a little Trigger step and couple other classes.

My raw implementation was to call self.build.startNextStep() in Trigger.run() method before any other actions in that method.

Is it interesting idea? Maybe I'm completelly wrong and there is already better way to achive this?

If idea is interesting, how can I describe it better? Text file with bigger description, or maybe pull request with code?
Now I have all changes done in form of subclasses, because it was easier for me to develop and test with real projects.
I did also changes in Build class to implement this feature.

Also I did monkey patch on StepsConnectorComponent.addStep method to provide safer adding steps.
There was cases when it was called simultaniosly from two places - from normal startNextStep() call and from startNextStep() in Trigger

**Comment by tardyp** [comment]
Hi @MikeAlexeev,
I like the idea a lot. This is a feature that was requested by my users at work as well.

This gives the problem of when do we send the "finished" event to reporters?
- at the end of the last step of the build?
   people want to get the mail that the release is available before the last long run test is finished.
- when all the steps, including not waited for triggered builds?
  people want to get mail when all the long run tests are finished.

Probably we would need another build state in order to implement that.

I think as this is a feature deeply inserted into the core, a PR would be nice with PoC implementation and integration tests in order to discuss how it should work. Then we can work on the corner cases unit tests.

**Comment by tardyp** [comment]
Here is a typical pipeline I can see for a release job that I use in my team (simplified):

- TriggerStep(waitForFinish=True, buidlers=['buildosx", "buildlinux"])  (those runs on big build machines)

- TriggerStep(waitForFinish=True, buidlers=['smoketestsosx", "smoketestslinux"]) (those run on machines with additional environment required by tests)

- Validation team wants to get a email at that point to start manual tests that are not yet automated, or not yet integrated in buildbot.

- TriggerStep(waitForFinish=False, buidlers=['longtestsosx", "longttestslinux"]) (those run maybe for one or two days)

- release team wants to get an email at that point to send notification to the customers.

Maybe you don't have the same process, and we are not talking about the same issues, then please also write description on the pipeline you want to describe

**Comment by MikeAlexeev** [comment]
In this case I think that one more builder can be added

For example:    
-  TriggerStep(waitForFinish=True, buidlers=['buildosx", "buildlinux"])
- TriggerStep(waitForFinish=True, buidlers=['smoketestsosx", "smoketestslinux"]) (those run on machines with additional environment required by tests)
- TriggerStep(waitForFinish=False, buidlers=["long_run_tasks"])
    - TriggerStep(waitForFinish=True, builders=["longtestsosx", "longtestslinux"])
    - send mail to release team
- send mail right after smoketests done as we don't wait for log_tun_tasks here

It can be implemented now without any changes. 
Maybe I'm wrong and didn't understand you right. Correct me please if it is so.

On my job we are transfering projects from other release system (it is inhouse development).
So structure of builders is done similar to structure on that system and can be changed a lot.

We use pairs (toolchain, component name) in builder names for all needed combinations of components and toolchains.

- Main builder (with force sched and parameters in UI)
    - TriggerStep(builders=['donwload deps']) 
    - TriggerStep(builders=['toolchain A']) 
        - TriggerStep(builders=['component A1']) 
        - TriggerStep(builders=['component A2'])
        - TriggerStep(builders=['component A3'])
    - TriggerStep(builders=['toolchain B'])
        - TriggerStep(builders=['component B1'])
        - TriggerStep(builders=['component B2'])
        - TriggerStep(builders=['component B3'])
    - TriggerStep(builders=['toolchain C'])
        - TriggerStep(builders=['component C4'])
        - TriggerStep(builders=['component C5'])
    - TriggerStep(builders=['toolchain D'])
        - TriggerStep(builders=['component D4'])
        - TriggerStep(builders=['component D5'])

    - TriggerStep(builders=['package rpm'])
    - TriggerStep(builders=['package deb'])
    - TriggerStep(builders=['package bsd'])


We use TriggerStep to start only one builder, so we have better control with doStepIf callables in TriggerSteps

For example build all with toolchains A,B and component name 3

All trigger steps are with waitForFinish=True, but I have added custom behavior to triggers so components can run simultaneously while parent build continue to wait child status.
It helps to parallel build process.

Also some components have to wait other components finished, i.e. they depend from other.
And they can start only if other components built and their status is ok.

I have PoC of such behavior too but this is another PR.

**Comment by staticfloat** [comment]
Did this ever get merged?  I have a couple of use cases for this functionality; it would allow for much more flexible parallel build stages (such as a "clean cache on all workers" meta-job).

**Comment by tardyp** [comment]
I don't remember this to have been merged.

**Comment by vit9696** [comment]
Hello, we had such a change implemented for a legacy buildbot a while ago, which I have just forwardported in https://github.com/vit9696/buildbot/commit/7caeaef3b514b7551c4fc917397db04d24c69498 and am currently testing. @tardyp would you be interested in merging it upstream?

**Comment by vit9696** [comment]
We ended implementing this feature with a custom trigger step without patching buildbot. Slightly simplified code in question looks as follows:

```py
class AsyncTrigger(steps.Trigger):
    def setAsyncLock(self, lock):
        self.asyncLock = lock

    @defer.inlineCallbacks
    def _createStep(self):
        """
        We need to generate step number with locks as this code
        may now run in parallel.
        """
        self.name = yield self.build.render(self.name)
        self.build.setUniqueStepName(self)
        self.stepid, self.number, self.name = yield self.master.data.updates.addStep(
            buildid=self.build.buildid,
            name=bytes2unicode(self.name))

    @defer.inlineCallbacks
    def addStep(self):
        """
        Create and start the step, noting that the name may be altered to
        ensure uniqueness.
        """
        yield self.asyncLock.run(self._createStep)
        yield self.master.data.updates.startStep(self.stepid)

class MyBuild(build.Build):
    def setupBuild(self):
        """
        Remember async locks and create an async lock itself,
        providing it to the async triggers.
        """
        super().setupBuild()

        self.asyncSteps = []
        self.asyncLock = defer.DeferredLock()

        for step in self.steps:
            if isinstance(step, AsyncTrigger):
                self.asyncSteps.append(step)
                step.setAsyncLock(self.asyncLock)

    def stopBuild(self, reason="<no reason given>", results=CANCELLED):
        """
        Interrupt not just the current step but also all async steps.
        """
        log.msg(f" {self}: stopping my build: {reason} {results}")
        if self.finished:
            return

        self.stopped = True

        for step in self.asyncSteps:
            step.interrupt(reason)

        return super().stopBuild(reason, results)

    def startNextStep(self):
        """
        Start more than one step when there are async steps in the build.
        """
        while True:
            try:
                s = self.getNextStep()
            except StopIteration:
                s = None
            if s:
                self.executedSteps.append(s)
                self.currentStep = s

                # Run all async steps we have right now.
                if isinstance(s, AsyncTrigger):
                    self._start_next_step_impl(s)
                    continue

                self._start_next_step_impl(s)
                return defer.succeed(None)

            if self.asyncSteps:
                return defer.succeed(None)

            return self.allStepsDone()

    def stepDone(self, results, step):
        """
        Remove step from asyncSteps to ensure last async step
        triggers allStepsDone().
        """
        if isinstance(step, AsyncTrigger):
            self.asyncSteps.remove(step)

        return super().stepDone(results, step)


class MyBuildFactory(factory.BuildFactory):
    def __init__(self, steps=None):
        """
        Use custom build factory to permit wrapping build methods.
        """
        super().__init__(steps)
        self.buildClass = MyBuild
```

**Comment by pktpls** [comment]
I'm very interested in this feature - we build the same codebase for ~30 architectures and want to parameterize/parallelize this (it's an OpenWrt-related project).

@vit9696 is this the code you are successfully running? It looks good but I'm getting `'AsyncTrigger' object has no attribute 'asyncLock'`, so it seems like `MyBuild.setupBuild()` hasn't executed yet when `AsyncTrigger.addStep()` is called. Buildbot 3.6.1 / Twisted 22.10.0

**Comment by pktpls** [comment]
It works nicely now :-) I'm using it together with `addStepsAfterCurrentStep()` and so I had to override that and call `setAsyncLock()` there too.

@vit9696 would you be okay with me adding this to our codebase at https://github.com/freifunk-berlin/falter-packages/?

**Comment by vit9696** [comment]
Hi, we made minor alterations to the code above, but in general it is about the same. Will try to look it up later. I think there were some issues when stopping tasks.

Feel free to use the code. It would be nice if you let me know about bugfixes you make though (if any) :-)

**Comment by pktpls** [comment]
We've been using and adapting @vit9696's code and it's pretty nice, thanks a lot! You can read our version here: https://github.com/freifunk-berlin/ansible/blob/main/roles/buildbot/files/asyncbuild.py

I made a few small changes/additions:
- Use `addStepsAfterCurrentStep()` to build an AsyncBuildGenerator based on shell command output. (We use these together with virtual builders.)
- Wait until all consecutive TriggerSteps are finished before running the next non-TriggerStep. Also wait before finishing the build.
- Add counter properties that let top-level build keep track of its triggered builds.

**Comment by congzhangzh** [comment]
@tardyp we add a concept project, which maybe help

https://github.com/congzhangzh/buildbot-parallel

later, we can make a pull request:)

**Comment by vit9696** [comment]
Hi @congzhangzh, sounds interesting, but your URL gives 404 error for me.

**Comment by congzhangzh** [comment]
> Hi @congzhangzh, sounds interesting, but your URL gives 404 error for me.

en，it’s private, once all work done, it will be public:)
